dbms/src/Interpreters/ExpressionAnalyzer.cpp

```diff
@@ -279,12 +279,14 @@ void ExpressionAnalyzer::tryMakeSetForIndexFromSubquery(const ASTPtr & subquery_
     SetPtr set = std::make_shared<Set>(settings.size_limits_for_set, true);
     set->setHeader(res.in->getHeader());
 
+    res.in->readPrefix();
```

Comment:
The function [TabSeparatedRowInputStream::readPrefix()](https://github.com/yandex/ClickHouse/blob/452a24efbb642ccbceb4d5bb7c3256312b76e043/dbms/src/Formats/TabSeparatedRowInputStream.cpp#L129) calls [setupAllColumnsByTableSchema()](https://github.com/yandex/ClickHouse/blob/452a24efbb642ccbceb4d5bb7c3256312b76e043/dbms/src/Formats/TabSeparatedRowInputStream.cpp#L72) which initializes the vector `column_indexes_for_input_fields`, which is used in the [read()](https://github.com/yandex/ClickHouse/blob/452a24efbb642ccbceb4d5bb7c3256312b76e043/dbms/src/Formats/TabSeparatedRowInputStream.cpp#L180) function to iterate through the columns.
Without calling `readPrefix()` here [TabSeparatedRowInputStream::read()](https://github.com/yandex/ClickHouse/blob/452a24efbb642ccbceb4d5bb7c3256312b76e043/dbms/src/Formats/TabSeparatedRowInputStream.cpp#L180) iterates through an empty list of columns without moving the file pointer and assigns all columns by default, then do it again and again in the infinite loop.